core/services/ardupilot_manager/firmware_download/FirmwareDownload.py

```diff
@@ -0,0 +1,206 @@
+import enum
+import json
+import pathlib
+from urllib.request import urlopen, urlretrieve
+
+from warnings import warn
+
+from packaging.version import Version
+from typing import Any, Dict, List, Optional
+
+
+class Vehicle(enum.Enum):
+    """Valid vehicle types to download"""
+
+    Sub = 1
+    Rover = 2
+    Plane = 3
+    Copter = 4
+
+
+class FirmwareDownload:
+    _manifest_remote = "https://firmware.ardupilot.org/manifest.json.gz"
+    _supported_firmware_format = "apj"
+
+    def __init__(self) -> None:
+        self._manifest: Dict[str, Any] = {}
+
+    @staticmethod
+    def _generate_random_filename(length: int = 16) -> pathlib.Path:
+        """Generate a random name for a temporary file.
+
+        Args:
+            length (int, optional): Size of the name. Defaults to 16.
+
+        Returns:
+            pathlib.Path: Path of the temporary file.
+        """
+        import random
```

Comment:
what is the reasoning for these local imports? this seems pretty non-standard.

Reply:
It has the imports that are only used inside this scope to track the imports easily, otherwise it makes harder to detect when a include is not being used. But since we have pylint, this is not necessary anymore.